net/src/ipv4/prefix.rs

```diff
@@ -0,0 +1,665 @@
+    fn eq(&self, other: &u8) -> bool {
+        self.0 == *other
+    }
+}
+
+impl Ipv4Prefix {
+    /// The root [`Ipv4Prefix`], aka 0.0.0.0/0
+    pub const ROOT: Ipv4Prefix = Ipv4Prefix::new_assert([0, 0, 0, 0], 0);
+
+    /// Validating a constructor which panics if the arguments are invalid.
+    /// Useful in const contexts and testing.
+    ///
+    /// Avoid this method outside const contexts or testing settings.
+    ///
+    /// # Panics
+    ///
+    /// * Panics if the provided prefix is greater than 32
+    /// * Panics if the provided address contains non-network bits.
+    #[must_use]
+    pub const fn new_assert(addr: [u8; 4], prefix: u8) -> Self {
+        let addr = Ipv4Addr::new(addr[0], addr[1], addr[2], addr[3]);
+        let prefix = Ipv4PrefixLen::new_assert(prefix);
+        let mask = (!0u32)
+            .overflowing_shl((Ipv4PrefixLen::MAX_LEN - prefix.0) as u32)
+            .0;
+        assert!(
+            addr.to_bits() & mask == addr.to_bits(),
+            "Ipv4Network address contains non network bits"
+        );
+        Ipv4Prefix(Ipv4Net::new_assert(addr, prefix.0))
+    }
+
+    /// Constructor which validates the arguments provided.
+    ///
+    /// # Errors
+    ///
+    /// * Returns [`InvalidIpv4Network::InvalidPrefix`] if the provided prefix length is greater
+    ///   than [`Ipv4PrefixLen::MAX_LEN`].
+    /// * Returns [`InvalidIpv4Network::AddressContainsNonNetworkBits`] if the provided address
+    ///   contains non-network bits.
+    #[tracing::instrument(level = "trace")]
+    pub fn new_strict(
+        addr: impl Into<Ipv4Addr> + Debug,
+        prefix: impl TryInto<Ipv4PrefixLen, Error = InvalidIpv4PrefixLength> + Debug,
+    ) -> Result<Ipv4Prefix, InvalidIpv4Network> {
+        let addr = addr.into();
+        let prefix = prefix
+            .try_into()
+            .map_err(InvalidIpv4Network::InvalidPrefix)?;
+        if addr.to_bits()
+            & (!0u32)
+                .overflowing_shl(u32::from(Ipv4PrefixLen::MAX_LEN - prefix.0))
+                .0
+            != addr.to_bits()
+        {
+            return Err(InvalidIpv4Network::AddressContainsNonNetworkBits(
+                addr, prefix,
+            ));
+        }
+        Ok(Ipv4Prefix(Ipv4Net::new_assert(addr, prefix.0)))
+    }
+
+    /// Create an [`Ipv4Prefix`] even if the argument contains non-network bits.
+    ///
+    /// This method is useful in the cases that
+    ///
+    /// 1. You can't trust your routing input and still don't wish to reject the routes (ideally,
+    ///    this does not happen).
+    /// 2. You need to convert an interface address assignment into a route
+    ///
+    /// This method will log a warning if the provided address contains non-network bits.
+    ///
+    /// # Errors
+    ///
+    /// * Returns an error if the provided prefix length is greater than [`Ipv4PrefixLen::MAX_LEN`].
+    #[tracing::instrument(level = "debug")]
+    pub fn new_tolerant<E>(
```

Comment:
I see the merit of this vs new_strict(). However, shouldn't we just agree on what's acceptable and have a single fallible constructor?

Reply:
I am happy to rename the method to `new` if that reduces syntax noise, but the other constructor is still quite useful in other settings.

It is one thing to accept invalid input and hammer it into "validity" by squinting at it and hoping you know what it means.

It is quite another to ignore invariants _within_ our own data structures.

For example, I think it is sorta kinda ok to accept a route with the host bits set under the theory that you know the prefix length and you are just assuming that the data source forgot to clear the low order bits or something.

It is another thing entirely to fail to honor invariants within your own data structures.  As we have discussed before, there isn't really any way to recover from internal programmer error.